pkg/service/storage/ucan.go

```diff
@@ -172,6 +179,100 @@ func NewUCANServer(storageService Service, options ...server.Option) (server.Ser
 				},
 			),
 		),
+		server.WithServiceMethod(
+			replica.AllocateAbility,
+			server.Provide(
+				replica.Allocate,
+				func(cap ucan.Capability[replica.AllocateCaveats], inv invocation.Invocation, iCtx server.InvocationContext) (replica.AllocateOk, fx.Effects, error) {
+					//
+					// UCAN Validation
+					//
+
+					// only service principal can perform an allocation
+					if cap.With() != iCtx.ID().DID().String() {
+						return replica.AllocateOk{}, nil, NewUnsupportedCapabilityError(cap)
+					}
+
+					//
+					// end UCAN Validation
+					//
+
+					// create the transfer invocation: an fx of the allocate invocation receipt.
+					trnsfInv, err := replica.Transfer.Invoke(
+						storageService.ID(),
+						storageService.ID(),
+						storageService.ID().DID().GoString(),
+						replica.TransferCaveats{
+							Space:    cap.Nb().Space,
+							Blob:     cap.Nb().Blob,
+							Location: cap.Nb().Location,
+							Cause:    inv.Link(),
+						},
+					)
+					if err != nil {
+						return replica.AllocateOk{}, nil, failure.FromError(err)
+					}
+					for block, err := range inv.Blocks() {
+						if err != nil {
+							return replica.AllocateOk{}, nil, fmt.Errorf("iterating replica allocate invocation blocks: %w", err)
+						}
+						if err := trnsfInv.Attach(block); err != nil {
+							return replica.AllocateOk{}, nil, fmt.Errorf("failed to replica allocate invocation block (%s) to transfer invocation: %w", block.Link().String(), err)
+						}
+					}
+
+					// read the location claim from this invocation to obtain the DID of the URL
+					// to replicate from on the primary storage node.
+					br, err := blockstore.NewBlockReader(blockstore.WithBlocksIterator(inv.Blocks()))
+					if err != nil {
+						return replica.AllocateOk{}, nil, failure.FromError(err)
+					}
+					claim, err := delegation.NewDelegationView(cap.Nb().Location, br)
+					if err != nil {
+						return replica.AllocateOk{}, nil, failure.FromError(err)
+					}
+
+					// TODO since there is a slice of capabilities here we need to validate the 0th is the correct one
+					// unsure what `With()` should be compared with for a capability.
+					lc, err := assert.LocationCaveatsReader.Read(claim.Capabilities()[0].Nb())
+					if err != nil {
+						return replica.AllocateOk{}, nil, failure.FromError(err)
+					}
+
+					if len(lc.Location) < 1 {
+						return replica.AllocateOk{}, nil, failure.FromError(fmt.Errorf("location missing from location claim"))
+					}
+
+					// TODO: which one do we pick if > 1?
+					replicaAddress := lc.Location[0]
+
+					// FIXME: use a real context, requires changes to server
+					ctx := context.TODO()
+					allocateResp, err := storageService.Capabilities().BlobAllocate(ctx, &capabilities.BlobAllocateRequest{
+						Space: cap.Nb().Space,
+						Blob:  cap.Nb().Blob,
+						Cause: inv.Link(),
```

Comment:
The replica/allocate is the cause of this blob/allocate invocation, yeah? Rather than the transfer?

Reply:
no I think cause here is the original space/blob/replicate which should be the Cause value in `cap.Nb().Cause`

Reply:
cause you are not ACTUALLY invoking blob/allocate here, you're just using a common handler

(I might rename the function `AllocateBlob` rather than `BlobAllocate` to make it more clear it's a command, not a capability)

Reply:
The AllocationStore called internally by `blobhandler.Allocate` expects a cause link for the `Put` operation. `inv.Link` is the `blob/replica/allocate` invocation, which is technically the cause(?) where `space/blob/replicate` is the cause of `blob/replica/allocate`, right?